### ISSUE TITLE
Support for __debug__ special variable

### ISSUE THREAD

**Issue by prusnak** [microsoft/pyright#630]
**Describe the bug**

the following code will fail to type-check:

```python
if __debug__:
    some_var = True

other_var = False

if __debug__:
    print(some_var == other_var)
```
 
error is:

```
  7:11 - error: "some_var" is possibly unbound (reportGeneralTypeIssues)
```

This is reasonable for plain old boolean variables. But pyright should know that `__debug__` is statically known to be `True` (maybe even configurable to `False`?) and evaluate the above code accordingly.

**VS Code extension or command-line**

command-line: pyright 1.1.33

**Comment by erictraut** [comment]
I don't think a static type checker should assume that `__debug__` is set to either `False` or `True`. It can be set differently depending on how the Python interpreter is started. Code should be written such that it runs in both cases.

**Comment by prusnak** [comment]
> Code should be written such that it runs in both cases.

It does. If `__debug__` is False, the code boils down to just `other_var = False`

**Comment by erictraut** [comment]
Yes, but a static type checker can't determine that for purposes of code flow analysis. Consider if you replaced `__debug__` with some other expression like `is_logging_enabled()`.

Pyright does special-case a few expressions whose value can be determined statically. These include `sys.version_info` and `sys.platform`. I think these are reasonable cases for the user to override because it's legitimate to write code that targets a particular version of Python or a particular platform. I don't think it's legitimate to write code that runs only in debug mode, so it wouldn't make sense to expose a configuration flag that tells the type checker to assume `__debug__` is always `True` or always `False`.

Here are a couple of workarounds:
1. Avoid conditional binding of variables. Always give a variable a defined value along all code paths. (`some_var = True`).
2. Suppress the type checking error using a `# type: none` comment.

**Comment by prusnak** [comment]
I will explain why this feature is interesting for us (and possibly the others) and especially why the following is not always true:

> I don't think it's legitimate to write code that runs only in debug mode

In our project, we use MicroPython - a Python3 implementation suited for microcontrollers. Among other things, it can "freeze" a Python module into a C code. While doing so, it removes all unreachable code making the resulting bytecode smaller. By simply using `-O1` or `-O0` we can easily produce two versions of the resulting binary: one with and the other without the debugging functionality. If a function call is used instead of `__debug__` directly, the code is not removed.

**Comment by enjoysmath** [comment]
@prusnak how do you pass in these flags using Thonny?

**Comment by reasonablecow** [comment]
The proposed solution with defineConstant does not work (pyright 1.1.359).
I compared the behavior with other variables e.g. __name__ (where it works just fine) and it seems that since __debug__ is a built-in constant (I can not assign to it in the code either), it does not work with defineConstant.
